### PR TITLE
Add support for funnel shifts + generalize ternary operators

### PR DESCRIPTION
**Pull request by pranavk**

### CODE REVIEW

**Comment by regehr** [comment]
this looks fine to me but I want @zhengyangl to review the changes since he touched this part of the synthesis code most recently

File: lib/Infer/ExhaustiveSynthesis.cpp [hunk 1]
```diff
@@ -322,71 +326,82 @@ void getGuesses(std::vector<Inst *> &Guesses,
   // need two reserved per select instruction
   Inst *C2 = IC.getReservedConst();
   Comps.push_back(C2);
+  Inst *C3 = IC.getReservedConst();
```
Comment (zhengyangl) [inline lib/Infer/ExhaustiveSynthesis.cpp]: Why create a new constant here, we want to prune (ter-op c,c,c) as early as possible.

File: lib/Infer/ExhaustiveSynthesis.cpp [hunk 2]
```diff
@@ -322,71 +326,82 @@ void getGuesses(std::vector<Inst *> &Guesses,
...
+      if (I->K == Inst::ReservedInst && I != I1)
         continue;
-      if ((*J)->K == Inst::ReservedConst && (*J) != C2)
+
+      if (I->K == Inst::ReservedConst && I != C1)
         continue;
 
-      // Prune (select cond, x, x)
-      if (I == J)
+      // (select i1, c, c)
```
Comment (zhengyangl) [inline lib/Infer/ExhaustiveSynthesis.cpp]: I would change this comment into (select, c, x, x) here

File: lib/Infer/ExhaustiveSynthesis.cpp (hunk 1 above)
Comment (zhengyangl) [reply lib/Infer/ExhaustiveSynthesis.cpp]: I see, ignore this comment

File: lib/Infer/ExhaustiveSynthesis.cpp [hunk 3]
```diff
@@ -322,71 +326,82 @@ void getGuesses(std::vector<Inst *> &Guesses,
...
         } else {
-          V = L;
+          V2 = J;
         }
 
-        if (V->Width != 1)
-          continue;
+        for (auto K : Comps) {
+          if (K->K == Inst::ReservedInst && K != I3)
+            continue;
```
Comment (zhengyangl) [inline lib/Infer/ExhaustiveSynthesis.cpp]: Also check (K->K == ReservedConst) && K != C3 here

File: lib/Infer/ExhaustiveSynthesis.cpp [hunk 4]
```diff
@@ -322,71 +326,82 @@ void getGuesses(std::vector<Inst *> &Guesses,
...
 
+          // PRUNE: (select cond, x, x)
+          if (Op == Inst::Select && J == K)
+            continue;
 
-        auto SelectInst = IC.getInst(Inst::Select, Width, { V, V1, V2 });
+          // PRUNE: {fshl, fshr} c, c, c
+          if ((Op == Inst::FShl || Op == Inst::FShr) &&
```
Comment (zhengyangl) [inline lib/Infer/ExhaustiveSynthesis.cpp]: Please put this prune right after L374. This prune works for all ter-op.

File: lib/Infer/ExhaustiveSynthesis.cpp [hunk 5]
```diff
@@ -322,71 +326,82 @@ void getGuesses(std::vector<Inst *> &Guesses,
...
+            V3 = IC.createHole(Width);
+          } else {
+            V3 = K;
+          }
 
-        addGuess(SelectInst, LHSCost, PartialGuesses, TooExpensive);
+          if (V2->Width != V3->Width)
+            continue;
+
+          if (V2->Width != Width)
```
Comment (zhengyangl) [inline lib/Infer/ExhaustiveSynthesis.cpp]: This line can be hoisted to the outer loop

File: lib/Infer/ExhaustiveSynthesis.cpp [hunk 6]
```diff
@@ -322,71 +326,82 @@ void getGuesses(std::vector<Inst *> &Guesses,
...
       } else {
-        V2 = *J;
+        V1 = I;
       }
 
-      if (V1->Width != V2->Width)
-        continue;
-
-      if (V1->Width != Width)
+      if (Op == Inst::Select && V1->Width != 1)
```
Comment (zhengyangl) [inline lib/Infer/ExhaustiveSynthesis.cpp]: Should there be a width check for the operators other than select?

**Comment by zhengyangl** [comment]
Looks good! a few comments inline.

**Comment by regehr** [comment]
let me know when this is done

File: lib/Infer/ExhaustiveSynthesis.cpp (hunk 3 above)
Comment (pranavk) [reply lib/Infer/ExhaustiveSynthesis.cpp]: yes, good point. i forgot

File: lib/Infer/ExhaustiveSynthesis.cpp (hunk 6 above)
Comment (pranavk) [reply lib/Infer/ExhaustiveSynthesis.cpp]: yup. added now.

**Comment by pranavk** [comment]
@regehr @zhengyangl Done. Just added unit tests too which works.

**Comment by zhengyangl** [comment]
LGTM :)

**Comment by regehr** [comment]
thanks!

**Pull request merged**